crates/ruff_db/src/system/memory_fs.rs

```diff
@@ -238,32 +253,30 @@ impl MemoryFileSystem {
         normalized.into_utf8_path_buf()
     }
 
-    pub fn read_directory(
-        &self,
-        path: impl AsRef<SystemPath>,
-    ) -> Result<impl Iterator<Item = Result<DirectoryEntry>> + '_> {
+    pub fn read_directory(&self, path: impl AsRef<SystemPath>) -> Result<ReadDirectory> {
```

Comment:
I'm happy to revert this pack to returning an `impl`. I had to change it, because a first version of the walker had to store the current iteration state in a `struct` which doesn't work with `impl` (Unless you use `dyn`). I introduced the named type and thought I'll leave it, because it can be useful. 

It's also nice that the type now implements `Debug`